proposed/dynamic-component-loading.md

```diff
@@ -0,0 +1,110 @@
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+```csharp
+class Assembly
+{
+    public static Assembly LoadFileWithDependencies(string path);
+}
+```
+
+At its core this is similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json`. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+
+```csharp
+class AssemblyLoadContext
+{
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(
+        string assemblyPath,
+        AssemblyLoadContext fallbackContext,
+        bool enableUnloading);
+}
+```
+
+"Advanced" version which would return an ALC instance and not just the main assembly. Allows for additional changes to the `AssemblyLoadContext`, like registering an event handler to the `Resolving` event for example.  
+Also adds the ability to specify:
+* `fallbackContext` which is the `AssemblyLoadContext` to defer to when the assembly resolution is not possible in this current context. By default this is the `AssemblyLoadContext.Default` (the app's context). This allows for creating effectively parent-child relationships between the load contexts.
+* `enableUnloading` which will mark the newly created load context to track references and to trigger unload for the load context and all the assemblies in it when possible.
+
+## High-level description of the implementation
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication). Specifically `hostpolicy.dll` is the component which parses `.deps.json` during app startup. See [host-components](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md) for more details. If this functionality is required when running with a custom host, the said host would need to provide this functionality to the runtime.
+* If the component has `.runtimeconfig.json` and/or `.runtimeconfig.dev.json` it will only be used to verify runtime version and provide probing paths.
```

Comment:
I'm sorry - probably missing the point of your question.
`.runtimeconfig.json` serves different purpose than `.deps.json`.  
The former is used to locate the runtime and the runtime options to start the runtime with. It also stores framework dependencies - this is somewhat weird since the "dependencies" part of that should be in `.deps.json` logically. On the other hand, the way to find the runtime is to locate the lowest-level framework, so that part needs to be in `.runtimeconfig.json`.
In a way the runtime-config part is meant to be human editted. The framework dependencies are generally machine generated as those fall out of the compilation process.  
The latter is used to locate specific assembly dependencies. This almost entirely a result of the compilation process and thus machine generated.  

Please note that vast majority of dynamically loaded components will not have `.runtimeconfig.json`. The proposed new functionality is about finding assembly dependencies and as such it's not really tied to the `.runtimeconfig.json` (other than the probing paths), on the other hand it is very much tied to `.deps.json`